src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala

```diff
@@ -412,6 +412,31 @@ class ResourceService(
     eventualSubjects.toList.sequence.map(_.toMap)
   }
 
+  //Disallow orphaning, leaving when it's via a group, or leaving when the resource is public
+  def leaveResource(resourceType: ResourceType, resourceId: FullyQualifiedResourceId, samUser: SamUser, samRequestContext: SamRequestContext): IO[List[Boolean]] = {
+    accessPolicyDAO.listAccessPolicies(resourceId, samRequestContext) flatMap { policiesForResource =>
+      //1. Make sure that there are no public policies for this resource. Leaving a public resource is not supported.
+      val isPublic = policiesForResource.exists(_.public)
+      if(isPublic) throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, "You may not leave a public resource."))
+
+      val policiesForUser = policiesForResource.filter(_.members.contains(samUser.id)).toList
+
+      //2. Make sure that the user has direct access that can be removed. Note that if user has both direct and indrect access,
+      //   this will succeed in removing their direct access. Which is still useful functionality.
+      if(policiesForUser.isEmpty) throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, "You can only leave a resource that you have direct access to."))
+
+      //3. Make sure that removing the user will not orphan it, i.e. there must be at least one other owner on an "owning" policy
+      val ownerPolicies = policiesForUser.filter(_.roles.contains(resourceType.ownerRoleName))
+      val removalOrphansResource = ownerPolicies.exists(_.members.size <= 1)
```

Comment:
I don't know how Sam policies are structured, but it seems surprising that this checks `policiesForUser` to determine if another user is an owner.